### PR TITLE
Image stream force

### PR DESCRIPTION
**Pull request by SimonRichardson**
### Checklist

 - [ ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?
 - [ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?
 - [ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?
 - [ ] Do comments answer the question of why design decisions were made?

----

## Description of change

*Please replace with a description about why this change is needed, along with a description of what changed?*

## QA steps

*Please replace with how we can verify that the change works?*

```sh
QA steps here
```

## Documentation changes

*Please replace with any notes about how it affects current user workflow? CLI? API?*

## Bug reference

*Please add a link to any bugs that this change is related to.*

### CODE REVIEW

File: cmd/juju/commands/bootstrap.go [hunk 1]
```diff
@@ -42,6 +42,7 @@ import (
 	"github.com/juju/juju/environs/bootstrap"
 	"github.com/juju/juju/environs/config"
 	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/imagemetadata"
```
Comment (SimonRichardson) [inline cmd/juju/commands/bootstrap.go]: I'm not a fan of this. I think this should be in core.

File: environs/imagemetadata/urls.go [hunk 2]
```diff
@@ -18,13 +18,25 @@ func ImageMetadataURL(source, stream string) (string, error) {
...
+
+// ValidDefaultImageMetadataURL returns true if the source passed in, is a
+// default image metadata url.
+// Does it satisfy either UbuntuCloudImagesURL or JujuStreamsImagesURL
+func ValidDefaultImageMetadataURL(source string) bool {
+	return source == UbuntuCloudImagesURL || source == JujuStreamsImagesURL
+}
+
+// ValidReleaseStream returns true if the stream is a released stream.
+func ValidReleaseStream(stream string) bool {
```
Comment (manadart) [inline environs/imagemetadata/urls.go]: `IsReleasedStream`?

File: cmd/juju/commands/bootstrap.go [hunk 3]
```diff
@@ -1448,3 +1459,35 @@ func newStringIfNonEmpty(s string) *string {
 	}
 	return &s
 }
+
+// shouldOverrideSeriesForce attempts to work out if it's a requirement for
+// the operator to supply force.
+func shouldOverrideSeriesForce(bootstrapModel map[string]interface{}) bool {
```
Comment (manadart) [inline cmd/juju/commands/bootstrap.go]: `shouldImplyForce`?

**Comment by howbazaar** [comment]
Hey @SimonRichardson and @manadart,

I don't think this is the right approach at all.

My understanding of the problem was that if the model had the image stream set to "daily", then a new release that has a future start date is considered "supported".

For this we probably need both client and server side changes. We don't want to fake this with a force flag.

@SimonRichardson and I talked about having the Juju concept of supported series be something that is in the Juju codebase, and we keep the `github.com/juju/os` package as independent. The `juju/os` package can have it's own concept of supported, but from Juju's perspective, we want a list of series that the version of Juju has been tested against. Consider the situation we just had with focal and the systemd changes. It is easily concievable that we'd hit another issue like this in 22.04, and we don't neceessarily want Juju 2.8 to consider that series as 'good'.

Consider having an internal package in `juju/juju` that determins whether a series is supported, and we should take into consideration whether the model is using the daily image stream or not.

Does all that make sense?

**Comment by manadart** [comment]
@howbazaar, @SimonRichardson, I was doing some recon for this, looking at the various methods that tell us whether a series is supported for bootstrap, workloads etc.

It seems to me that if Juju supports Focal, then it supports Focal - why not just set the entry to `Supported: true` in `juju/os/series`? An image will only be found if the daily stream is used, which is the single gate that we require.

**Comment by SimonRichardson** [comment]
@howbazaar we all agree this was a bad way to do this, hence my long commit message :smile: 

Closed to prevent accidental merging.

Superseded: #11332

**Pull request closed without merging**